### PR TITLE
`Int` & `Long` backed `BitChunk`

### PR DESCRIPTION
**Pull request by 987Nabil**
- `BitChunk` versions for underlying `Chunk[Int]` / `Chunk[Long]` (#6849)
- Extract abstract class with common logic for BitChunk's

### CODE REVIEW

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 1]
```diff
@@ -1100,6 +1112,12 @@ sealed abstract class Chunk[+A] extends ChunkLike[A] { self =>
 
 object Chunk extends ChunkFactory with ChunkPlatformSpecific {
 
+  sealed trait Endianness
```
Comment (987Nabil) [inline core/shared/src/main/scala/zio/Chunk.scala:1115]: I am not sure, if this is a good place for it.

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 2]
```diff
@@ -1692,45 +1712,51 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
     }
   }
 
-  private[zio] final case class BitChunk(bytes: Chunk[Byte], minBitIndex: Int, maxBitIndex: Int)
-      extends Chunk[Boolean] {
+  private[zio] sealed abstract class BitChunk[T](
+    chunk: Chunk[T],
+    val x: Int,
+    val x2: Int
```
Comment (987Nabil) [inline core/shared/src/main/scala/zio/Chunk.scala:1718]: Maybe someone has a better idea for a name?

**Comment by adamgfraser** [comment]
Can you retarget this against the `series/2.x` branch?

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 3]
```diff
@@ -1692,64 +1712,66 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
...
+      val toTake = (index + x2) >> x
+      newBitChunk(chunk.take(toTake), minBitIndex, index)
     }
 
     override def foreach[A](f: Boolean => A): Unit = {
-      val minByteIndex    = (minBitIndex + 7) >> 3
-      val maxByteIndex    = maxBitIndex >> 3
-      val minFullBitIndex = (minByteIndex << 3) min maxBitIndex
-      val maxFullBitIndex = (maxByteIndex << 3) max minFullBitIndex
+      if (this.isEmpty) return
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/Chunk.scala:1747]: Would avoid using `return`.

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 4]
```diff
@@ -1692,64 +1712,66 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
     }
   }
 
-  private[zio] final case class BitChunk(bytes: Chunk[Byte], minBitIndex: Int, maxBitIndex: Int)
-      extends Chunk[Boolean] {
+  private[zio] sealed abstract class BitChunkBase[T](
+    chunk: Chunk[T],
+    val x: Int,
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/Chunk.scala:1717]: Isn't there really just one parameter here that is like `bits` and then `x` is `bits` and `x2` is `2^bits - 1`?

File: core/shared/src/main/scala/zio/Chunk.scala (hunk 3 above)
Comment (987Nabil) [reply core/shared/src/main/scala/zio/Chunk.scala:1747]: Actually, this was only for debugging and I forgot to remove it. So I'll do

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 5]
```diff
@@ -1692,64 +1712,66 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
...
         i += 1
       }
     }
 
-    override def take(n: Int): BitChunk = {
-      val index  = (minBitIndex + n) min maxBitIndex
-      val toTake = (index + 7) >> 3
-      BitChunk(bytes.take(toTake), minBitIndex, index)
-    }
+    protected def foreachOptimized[A](f: Boolean => A, elem: T): Unit
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/Chunk.scala:1769]: Maybe `foreachElement`?

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 6]
```diff
@@ -1762,9 +1784,180 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
 
     override private[zio] def reverseArrayIterator[A1 >: Boolean]: Iterator[Array[A1]] =
       arrayIterator
+  }
+
+  private[zio] final case class BitChunk(bytes: Chunk[Byte], minBitIndex: Int, maxBitIndex: Int)
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/Chunk.scala:1789]: Maybe this should be `BitChunkByte` and then `BitChunkBase` can just become `BitChunk`?

File: core/shared/src/main/scala/zio/Chunk.scala (hunk 1 above)
Comment (987Nabil) [reply core/shared/src/main/scala/zio/Chunk.scala:1115]: idk about that. `BitChunk` is the byte version. And the only one that does not care about `Endianness`.

**Review by adamgfraser — CHANGES_REQUESTED** [review]
Great job bringing some order to some very low level code! Made a couple of suggestions but this looks great!

File: core/shared/src/main/scala/zio/Chunk.scala (hunk 6 above)
Comment (987Nabil) [reply core/shared/src/main/scala/zio/Chunk.scala:1789]: I tried that and got some "Binary incompatibility" errors. But maybe that's ok, if it only goes in zio 2.x?

File: core/shared/src/main/scala/zio/Chunk.scala (hunk 1 above)
Comment (adamgfraser) [reply core/shared/src/main/scala/zio/Chunk.scala:1115]: See my comment below regarding potentially renaming `BitChunk` to `BitChunkByte`.

File: core/shared/src/main/scala/zio/Chunk.scala (hunk 6 above)
Comment (adamgfraser) [reply core/shared/src/main/scala/zio/Chunk.scala:1789]: Yes that is one of the advantages of targeting the `series/2.x` branch.

File: core/shared/src/main/scala/zio/Chunk.scala (hunk 4 above)
Comment (987Nabil) [reply core/shared/src/main/scala/zio/Chunk.scala:1717]: I found a solution, I guess.

**Review by adamgfraser — DISMISSED** [review]
Looks great! Had one minor suggestion about naming but otherwise looks good to go.

**Comment by adamgfraser** [comment]
@987Nabil Can you resolve conflicts?

File: core/shared/src/main/scala/zio/Chunk.scala [hunk 7]
```diff
@@ -92,11 +91,23 @@ sealed abstract class Chunk[+A] extends ChunkLike[A] with Serializable { self =>
...
+  /**
+   * Converts a chunk of longs to a chunk of bits.
+   */
+  final def asBitsLong(endianness: Chunk.BitChunk.Endianness)(implicit ev: A <:< Long): Chunk[Boolean] =
+    Chunk.BitChunkLong(self.asInstanceOf[Chunk[Long]], endianness, 0, length << 6)
+
   /**
    * Converts a chunk of bytes to a chunk of bits.
    */
   final def asBits(implicit ev: A <:< Byte): Chunk[Boolean] =
```
[the remark this answers is not in the document]
Comment (987Nabil) [reply core/shared/src/main/scala/zio/Chunk.scala:109]: Do we need to document this in the migration guide?

**Comment by 987Nabil** [comment]
@adamgfraser I am on it

**Comment by adamgfraser** [comment]
@987Nabil That would be great but can also be done in a follow up!

**Comment by 987Nabil** [comment]
@adamgfraser I left the bit ops only on `BitChunkByte`. Needs a follow up ticket I guess.

**Comment by 987Nabil** [comment]
@adamgfraser I think I found a small bug. I guess the class tag of any `BitChunk` has to be `ClassTag.Boolean`. See my last commit.

**Comment by adamgfraser** [comment]
@987Nabil That's right!

**Review by adamgfraser — APPROVED** [review]

**Pull request merged**